addons/website/models/website_page.py

```diff
@@ -229,6 +238,94 @@ def _set_cache_response(self, cache_key, response):
+            'base_domain': domain,
+            'requires_sudo': requires_sudo,
+            'search_fields': search_fields,
+            'fetch_fields': fetch_fields,
+            'mapping': mapping,
+            'icon': 'fa-file-o',
+        }
+
+    @api.model
+    def _search_fetch(self, search_detail, search, limit, order):
+        with_description = 'description' in search_detail['mapping']
+        results, count = super()._search_fetch(search_detail, search, limit, order)
+        if with_description and search:
+            # Perform search in translations
+            # TODO Remove when domains will support xml_translate fields
+            query = sql.SQL("""
+                SELECT {table}.{id}
+                FROM {table}
+                LEFT JOIN ir_ui_view v ON {table}.{view_id} = v.{id}
+                LEFT JOIN ir_translation t ON v.{id} = t.{res_id}
+                WHERE t.lang = {lang}
+                AND t.name = ANY({names})
+                AND t.type = 'model_terms'
+                AND t.value ilike {search}
+                LIMIT {limit}
+            """).format(
+                table=sql.Identifier(self._table),
+                id=sql.Identifier('id'),
+                view_id=sql.Identifier('view_id'),
+                res_id=sql.Identifier('res_id'),
+                lang=sql.Placeholder('lang'),
+                names=sql.Placeholder('names'),
+                search=sql.Placeholder('search'),
+                limit=sql.Placeholder('limit'),
+            )
+            self.env.cr.execute(query, {
+                'lang': self.env.lang,
+                'names': ['ir.ui.view,arch_db', 'ir.ui.view,name'],
+                'search': '%%%s%%' % escape_psql(search),
+                'limit': limit,
+            })
+            ids = {row[0] for row in self.env.cr.fetchall()}
+            ids.update(results.ids)
+            domains = search_detail['base_domain'].copy()
+            domains.append([('id', 'in', list(ids))])
+            domain = expression.AND(domains)
+            model = self.sudo() if search_detail.get('requires_sudo') else self
+            results = model.search(
+                domain,
+                limit=limit,
+                order=search_detail.get('order', order)
+            )
+            count = max(count, len(results))
+
+        def filter_page(search, page, all_pages):
+            if 'url' not in order and not all_pages._is_most_specific_page(page):
+                return False
+            if search:
+                text = '%s %s' % (page.name, page.url)
+                if with_description:
+                    text = '%s %s' % (text, text_from_html(page.arch))
+                pattern = '|'.join([re.escape(search_term) for search_term in search.split()])
+                return re.findall('(%s)' % pattern, text, flags=re.I) if pattern else False
+            return True
+        results = results.filtered(lambda result: filter_page(search, result, results))
```

Comment:
this filtering has quadratic complexity, since `all_pages._is_most_specific_page` will iterate over all results again. is that necessary?

Reply:
Could not actually get rid of the behavior itself => Replaced by a single loop on pages sorted by url and "specificity".